packages/@lwc/engine-core/src/framework/stylesheet.ts

```diff
@@ -138,16 +138,42 @@ export function getStylesheetsContent(vm: VM, template: Template): string[] {
     return content;
 }
 
+// It might be worth caching this to avoid doing the lookup repeatedly, but
+// perf testing has not shown it to be a huge improvement yet:
+// https://github.com/salesforce/lwc/pull/2460#discussion_r691208892
+function getNearestNativeShadowComponent(vm: VM): VM | null {
+    let owner: VM | null = vm;
+    while (!isNull(owner)) {
+        if (owner.renderMode === RenderMode.Shadow && owner.shadowMode === ShadowMode.Native) {
+            return owner;
+        }
+        owner = owner.owner;
+    }
+    return owner;
+}
+
 export function createStylesheet(vm: VM, stylesheets: string[]): VNode | null {
     const { renderer, renderMode, shadowMode } = vm;
     if (renderMode === RenderMode.Shadow && shadowMode === ShadowMode.Synthetic) {
         for (let i = 0; i < stylesheets.length; i++) {
             renderer.insertGlobalStylesheet(stylesheets[i]);
         }
-        return null;
-    } else {
-        // native shadow or light DOM
+    } else if (renderer.ssr) {
+        // native shadow or light DOM, SSR
         const combinedStylesheetContent = ArrayJoin.call(stylesheets, '\n');
         return createInlineStyleVNode(combinedStylesheetContent);
+    } else {
+        // native shadow or light DOM, DOM renderer
+        const root = getNearestNativeShadowComponent(vm);
+        const isGlobal = isNull(root);
+        for (let i = 0; i < stylesheets.length; i++) {
+            if (isGlobal) {
+                renderer.insertGlobalStylesheet(stylesheets[i]);
+            } else {
+                // local level
+                renderer.insertStylesheet(stylesheets[i], root!.cmpRoot);
+            }
+        }
```

Comment:
(for my personal knowledge) I wonder if the js engine optimizes this code, do you know @nolanlawson ?:

```suggestion
        if (isGlobal) {
             for (let i = 0; i < stylesheets.length; i++) {
                 renderer.insertGlobalStylesheet(stylesheets[i]);
             }
        } else {
            for (let i = 0; i < stylesheets.length; i++) {
                 // local level
                renderer.insertStylesheet(stylesheets[i], root!.cmpRoot);
             }
        }
```

Reply:
@jodarove I admit I don't know. The `isGlobal` is a constant, so I would hope the engine optimizes it?

Unfortunately my benchmark wouldn't capture it, because it only has one stylesheet per template.

Reply:
@jodarove The difference is pretty small (if there even is a difference – I seem to get different results in JSBench if I run multiple times). I'd say let's stick with the more readable version, especially given that most templates will have just one stylesheet.

![Screen Shot 2021-09-01 at 10 10 54 AM](https://user-images.githubusercontent.com/283842/131714444-e7f4ea88-a403-4414-bb83-43a8f934a901.png)